src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -4400,350 +4405,727 @@ internal static IEnumerable<CompletionResult> CompleteFilename(CompletionContext
+            if (resolvedPaths is null || resolvedPaths.Count == 0)
+            {
+                return CommandCompletion.EmptyCompletionResult;
+            }
+
+            var useRelativePath = context.GetOption("RelativePaths", @default: defaultRelativePath);
+            if (useRelativePath && providerSeparatorIndex != -1)
+            {
+                // User must have requested relative paths but that's not valid with provider paths.
+                return CommandCompletion.EmptyCompletionResult;
+            }
+
+            var resolvedProvider = ((PathInfo)resolvedPaths[0].BaseObject).Provider;
+            string providerPrefix;
+            if (providerSeparatorIndex == -1)
+            {
+                providerPrefix = string.Empty;
+            }
+            else if (providerSeparatorIndex == resolvedProvider.Name.Length)
+            {
+                providerPrefix = $"{resolvedProvider.Name}::";
+            }
+            else
+            {
+                providerPrefix = $"{resolvedProvider.ModuleName}\\{resolvedProvider.Name}::";
+            }
+
+            List<CompletionResult> results;
+            switch (resolvedProvider.Name)
+            {
+                case FileSystemProvider.ProviderName:
+                    results = GetFileSystemProviderResults(
+                        context,
+                        resolvedProvider,
+                        resolvedPaths,
+                        filter,
+                        extension,
+                        containerOnly,
+                        useRelativePath,
+                        useLiteralPath,
+                        inputUsedHomeChar,
+                        providerPrefix,
+                        stringType);
+                    break;
+
+                default:
+                    results = GetDefaultProviderResults(
+                        context,
+                        resolvedProvider,
+                        resolvedPaths,
+                        filter,
+                        containerOnly,
+                        useRelativePath,
+                        useLiteralPath,
+                        inputUsedHomeChar,
+                        providerPrefix,
+                        stringType);
+                    break;
+            }
+
+            return results.OrderBy(x => x.ToolTip);
+        }
+
+        private static List<CompletionResult> GetFileSystemProviderResults(
+            CompletionContext context,
+            ProviderInfo provider,
+            Collection<PSObject> resolvedPaths,
+            string filterText,
+            HashSet<string> includedExtensions,
+            bool containersOnly,
+            bool relativePaths,
+            bool literalPaths,
+            bool inputUsedHome,
+            string providerPrefix,
+            StringConstantType stringType)
+        {
+            var enumerationOptions = _enumerationOptions;
```

Comment:
Might be worth having a debug assert checking that the provider is the filesystemprovider